src/sentry/utils/functional.py

```diff
@@ -57,3 +57,20 @@ def compact(seq):
 
     elif isinstance(seq, list):
         return [k for k in seq if k is not None]
+
+
+def cached(cache, function, *args, **kwargs):
+    """Calls ``function`` or retrieves its return value from the ``cache``.
+
+    This is similar to ``functools.cache``, but uses a custom cache instead
+    of a global one. The cache can be shared between multiple functions.
+    """
+    key = (id(function), args, tuple(kwargs.items()))
```

Comment:
* you can hash the function itself, that should not be a problem. it's better because id can be reused if function is freed, but it isn't if it's referenced in cache
* sort order of kwargs.items() is unstable